src/algebra/hierarchy_design.lean

```diff
@@ -0,0 +1,182 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison, Eric Weiser
+-/
+import tactic.doc_commands
+
+/-!
+# Documentation of the algebraic hierarchy
+
+-/
+
+/--
+# The algebraic hierarchy
+
+In any theorem proving environment,
+there are difficult decisions surrounding the design of the "algebraic hierarchy".
+
+There is a danger of exponential explosion in the number of gadgets,
+especially once interactions between algebraic and order/topological/etc structures are considered.
+
+In mathlib, we try to avoid this by only introducing new algebraic typeclasses either
+1. when there is "real mathematics" to be done with them, or
+2. when there is a meaninful gain in simplicity by factoring out a common substructure.
+
+(As examples, at this point we don't have `loop`, or `unital_magma`,
+but we do have `lie_submodule` and `topological_field`!)
+
+Generally in mathlib we use the extension mechanism (so `comm_ring` extends `ring`)
+rather than mixins (e.g. with separate `ring` and `comm_mul` classes),
+in part because of the potential blow-up in term sizes described at
+https://www.ralfj.de/blog/2019/05/15/typeclasses-exponential-blowup.html
+However there is tension here, as it results in considerable duplication in the API,
+particularly in the interaction with order structures.
+
+This library note is not intended as a design document
+justifying and explaining the history of mathlib's algebraic hierarchy!
+Instead it is intended as a developer's guide, for contributors wanting to extend
+(either new leaves, or new intermediate classes) the algebraic hierarchy as it exists.
+
+(Ideally we would have both a tour guide to the existing hierarchy,
+and an account of the design choices.
+See https://arxiv.org/abs/1910.09336 for an overview of mathlib as a whole,
+with some attention to the algebraic hierarchy and
+https://leanprover-community.github.io/mathlib-overview.html
+for a summary of what is in mathlib today.)
+
+## Instances
+
+When adding a new typeclass `Z` to the algebraic hierarchy
+one should attempt to add the following constructions and results,
+when applicable:
+
+* Instances tranferred elementwise to products, like `prod.monoid`.
+  See `algebra.group.prod` for more examples.
+  ```
+  instance prod.Z [Z M] [Z N] : Z (M × N) := ...
+  ```
+* Instances tranferred elementwise to pi types, like `pi.monoid`.
+  See `algebra.group.pi` for more examples.
+  ```
+  instance pi.Z [∀ i, Z $ f i] : Z (Π i : I, f i) := ...`
+  ```
+* Instances tranferred to `opposite M`, like `opposite.monoid`.
+  See `algebra.opposites` for more examples.
+  ```
+  instance opposite.Z [Z M] : Z (opposite M) := ...`
+  ```
```

Comment:
````suggestion
  ```
* Instances transferred to `ulift M`, like `ulift.monoid`.
  See `algebra.group.ulift` for more examples.
  ```
  instance ulift.Z [Z M] : Z (ulift M) := ...
  ```
````